arrow-core-data/src/main/kotlin/arrow/core/Option.kt

```diff
@@ -395,7 +418,137 @@ sealed class Option<out A> : OptionOf<A> {
 
     operator fun <A> invoke(a: A): Option<A> = Some(a)
 
+    inline fun <A> catch(recover: (Throwable) -> Unit, f: () -> A): Option<A> =
+      try {
+        just(f())
+      } catch (t: Throwable) {
+        raiseError(recover(t.nonFatalOrThrow()))
+      }
+
     fun <A> empty(): Option<A> = None
+
+    fun <A> raiseError(e: Unit): Option<A> = None
```

Comment:
I don't we should add this API for `Option` and otherwise I'd define it as `fun <A> raiseError(): Option<A> = None`.

IMO `just`, `raiseError` are counter intinuitive if not used in a polymorphic context. I'm unsure if `Option.just(1)` and `Either.just(1)` bring any additional value over `Option.some(1)` or `Either.right(1)`?

We need to review all these different constructors in all the data types before 0.12.0. I think for `Either` we have `Either.just(1)`, `Either.right(1)`, `1.right()`, `Either.raiseError("1")`, `Either.left("1")` and `"1".left()"`.
Futhermore Kotlin favors a type argument over returning `Nothing` like we currently do with `right()` and `left()` in `Either`. For example Kotlin Std: `emptyList<A>()` instead of `emptyList()` which would return `List<Nothing>`.

Reply:
I think we should get rid of just and raiseError. people will choose constructors over just and agreed that Option should not have a raiseError. For either extensions like .right(), .left() etc I think we should consolidate to use constructors or constructor like functions outside of the companion. I don't see any value to the dot syntax.or the companion factories beside confusing users with multiple ways to do the same.

Reply:
100% agreed.

Reply:
Makes sense, adding `raiseError` sounded weird to me but I put it because Option implemented `ApplicativeError` so I wasn't sure if we're ok removing that method.